llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp

```diff
@@ -3514,11 +3520,15 @@ void DwarfDebug::addDwarfTypeUnitType(DwarfCompileUnit &CU,
       return;
     }
 
-    // If the type wasn't dependent on fission addresses, finish adding the type
-    // and all its dependent types.
     for (auto &TU : TypeUnitsToAdd) {
       InfoHolder.computeSizeAndOffsetsForUnit(TU.first.get());
       InfoHolder.emitUnit(TU.first.get(), useSplitDwarf());
+      if (getDwarfVersion() >= 5 &&
+          getAccelTableKind() == AccelTableKind::Dwarf) {
+        addTypeUnit(std::move(TU.first));
```

Comment:
If I recall correctly... it's important that type units get discarded once emitted, for compiler memory use reasons. So keeping them around until we emit the name table might not be feasible.

Have you/can you measure the memory usage increase this change causes for some practical examples?

Reply:
> If I recall correctly... it's important that type units get discarded once emitted, for compiler memory use reasons. So keeping them around until we emit the name table might not be feasible.
> 
> Have you/can you measure the memory usage increase this change causes for some practical examples?

I have not. Were you thinking for one invocation of clang, or to build whole project?
What do you think a reasonable number of type units pre compilation modules are?

Reply:
Won't memory foot print still be less then when we do monolithic dwarf?

Reply:
It wouldn't be, no - type units have overhead necessary to make them standalone units, so the overhead is higher (& in any case, even if it isn't higher - it's still a lost opportunity/regression to substantially increase the memory overhead when using type units)

Reply:
> I have not. Were you thinking for one invocation of clang, or to build whole project?
> What do you think a reasonable number of type units pre compilation modules are?

Given I expect this to be a fairly clear non-starter, I'd probably test something pretty simple like, say, Clang's Sema.cpp - it's a bit of a big/rough thing to compile, probably has enough types to show up a significant regression I'd expect would be caused by keeping all the TUs alive to the end of the compilation.

Reply:
> > I have not. Were you thinking for one invocation of clang, or to build whole project?
> > What do you think a reasonable number of type units pre compilation modules are?
> 
> Given I expect this to be a fairly clear non-starter, I'd probably test something pretty simple like, say, Clang's Sema.cpp - it's a bit of a big/rough thing to compile, probably has enough types to show up a significant regression I'd expect would be caused by keeping all the TUs alive to the end of the compilation.

ok, let me do some memory overhead testing, and also seperate this in to two changes.

Reply:
> > I have not. Were you thinking for one invocation of clang, or to build whole project?
> > What do you think a reasonable number of type units pre compilation modules are?
> 
> Given I expect this to be a fairly clear non-starter, I'd probably test something pretty simple like, say, Clang's Sema.cpp - it's a bit of a big/rough thing to compile, probably has enough types to show up a significant regression I'd expect would be caused by keeping all the TUs alive to the end of the compilation.

Generated response file from command in compile_commands.json for Debug Build and used release build of clang to compile Sema.cpp
In the resulting object file there are 22124 .debug_info sections, so 22123 Type Units.

alias time="$(which time) -f '\t%E real,\t%U user,\t%S sys,\t%K amem,\t%M mmem'"
time ~/local/llvm-build-upstream-release/bin/clang++ @response.txt

with changes .debug_names generated
0:15.85 real,   15.41 user,     0.42 sys,       0 amem, 658136 mmem
0:15.15 real,   14.74 user,     0.39 sys,       0 amem, 658460 mmem
0:15.46 real,   15.05 user,     0.39 sys,       0 amem, 658444 mmem

without changes current default behavior .debug_names not generated 
0:15.94 real,   15.52 user,     0.41 sys,       0 amem, 658792 mmem
0:15.25 real,   14.81 user,     0.43 sys,       0 amem, 658452 mmem
0:15.58 real,   15.11 user,     0.46 sys,       0 amem, 658592 mmem

Is this what you were looking for?

Reply:
Yep, though it surprises me - do you have the full command line & preprocessed file? (you could add a pragma sema crash to the file to get clang to spit out a reproducer script/file - so I could try to reproduce this)

Reply:
[response.txt](https://github.com/llvm/llvm-project/files/12835121/response.txt)
[Sema-7c717f.cpp.txt](https://github.com/llvm/llvm-project/files/12835125/Sema-7c717f.cpp.txt)
[Sema-7c717f.sh.txt](https://github.com/llvm/llvm-project/files/12835132/Sema-7c717f.sh.txt)

Reply:
Hopefully I didn't miss something obvious.
[response.txt](https://github.com/llvm/llvm-project/files/12835417/response.txt)
[Sema-f715f6.cpp.txt](https://github.com/llvm/llvm-project/files/12835418/Sema-f715f6.cpp.txt)
[Sema-f715f6.sh.txt](https://github.com/llvm/llvm-project/files/12835427/Sema-f715f6.sh.txt)

Reply:
ran valgrind --tool=massif
https://valgrind.org/docs/manual/ms-manual.html#ms-manual.using-print
With Names:
![image](https://github.com/llvm/llvm-project/assets/43973793/c4f5debe-a0e3-4c47-9bf1-6b7c2b88b97f)
Without Names:
![image](https://github.com/llvm/llvm-project/assets/43973793/7f095780-5404-4b0e-b8b0-22d04464e02b)

Reply:
OK using LLC
time ~/local/llvm-build-upstream-release/bin/llc -filetype=obj tools/clang/lib/Sema/CMakeFiles/obj.clangSema.dir/Sema.cpp.ir.o -dwarf-version=5 -generate-type-units -mtriple x86_64-pc-linux -o foo.o
Looks like there is some reduction.
with changes
0:07.42 real,   7.21 user,      0.19 sys,       0 amem, 335256 mmem
0:07.96 real,   7.76 user,      0.19 sys,       0 amem, 334812 mmem
0:07.55 real,   7.35 user,      0.19 sys,       0 amem, 334112 mmem
 without changes 
 0:07.19 real,   7.01 user,      0.16 sys,       0 amem, 307616 mmem
0:07.27 real,   7.10 user,      0.17 sys,       0 amem, 305080 mmem
0:07.21 real,   7.05 user,      0.15 sys,       0 amem, 303628 mmem

So about 35MB difference.

Reply:
Yeah, that seems significant enough that it'd be worth preserving the behavior if reasonably possible.

So this work might end up looking more like a patch to change the way the current index works - to not depend on DIEs having a long lifetime (replacing DIE pointers with offsets - I guess maybe a union of the two, and we go through some lowering phase where the DIE pointer is read, the offset is retrieved (once known) and then the DIE pointer is replaced with the offset value) - just a pure refactor, no functional change. Then this patch could benefit from that restructure & avoid the memory regression/wouldn't need to extend the lifetime of type units.

Sound good?

Reply:
Yeah something like that. As we briefly discussed during llvm dev meeting just carry relative offset and tag info to accelerator table.
I will start looking into it on Monday, and once that is done I'll post a new PR. This time I know what button not-to push to merge it by mistake. lol

For LLD patch talking to @MaskRay seems like I was on a right path trying to add more logic to InputSection::relocateNonAlloc. Although I am still not clear how to figure out if TU was comdat from a Symbol.

Reply:
Hmm - I still wouldn't've thought you'd need to check if the TU is a comdat. That there'd already be existing code that's resolving/tombstoning - and it's just a question of picking a different tombstone if the relocation happens to be in the .debug_info section. Nothing about checking if it's comdat, or other details.

Reply:
Maybe I am not looking in the right place. For relocations in .debug_names into .debug_info it doesn't hit any of the tombstone code in that function. I'll circle back to it, after llvm part.

Reply:
Ah, perhaps there's a different codepath since the debug info section isn't SHF_ALLOC. Sorry, might be more involved.